pkg/util/testutil/docker/run.go

```diff
@@ -0,0 +1,108 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build test
+
+// Package docker provides API to manage docker/docker-compose lifecycle in UTs
+package docker
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/testutil"
+)
+
+// Run starts the container/s and ensures their successful invocation
+// LifecycleConfig is an interface that abstracts the configuration of the container/s
+// Use NewRunConfig to run a single docker container or NewComposeConfig to spin docker-compose
+// This method is using testing.TB interface to handle cleanup and logging during UTs execution
+func Run(t testing.TB, cfg LifecycleConfig) error {
+	var err error
+	var ctx context.Context
+	for i := 0; i < cfg.Retries(); i++ {
+		t.Helper()
+		// Ensuring no previous instances exists.
+		killPreviousInstances(cfg)
+
+		scanner := testutil.NewScanner(cfg.LogPattern(), make(chan struct{}, 1))
+		// attempt to start the container/s
+		ctx, err = run(t, cfg, scanner)
+		if err != nil {
+			t.Logf("could not start %s: %v", cfg.Name(), err)
+			//this iteration failed, retry
+			continue
+		}
+
+		//check container logs for successful start
+		if err = checkReadiness(ctx, cfg, scanner); err == nil {
+			// target container/s started successfully, we can stop the retries loop and finish here
+			return nil
+		}
+		scanner.PrintLogs(t)
+		t.Logf("failed to start %s server, retrying: %v", cfg.Name(), err)
+		time.Sleep(5 * time.Second)
+	}
+	return err
+}
+
+// we do best-effort to kill previous instances, hence ignoring any errors
+func killPreviousInstances(cfg LifecycleConfig) {
+	// Ensuring the following command won't block forever
+	timedContext, cancel := context.WithTimeout(context.Background(), cfg.Timeout())
+	defer cancel()
+	args := cfg.commandArgs(kill)
+
+	// Ensuring no previous instances exists.
+	c := exec.CommandContext(timedContext, "docker", args...)
+	c.Env = append(c.Env, cfg.Env()...)
+
+	// run synchronously to ensure all instances are killed
+	_ = c.Run()
```

Comment:
Running it asynchronously can yield side effects, as killing the instance we've just launched
In our CI, I suspect it is going to generate a couple of flaky tests